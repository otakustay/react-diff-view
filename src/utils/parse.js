import parser from 'gitdiff-parser';

const zipChanges = changes => {
    const [result] = changes.reduce(
        ([result, last, lastDeletionIndex], current, i) => {
            if (!last) {
                result.push(current);
                return [result, current, current.isDelete ? i : -1];
            }

            if (current.isInsert && lastDeletionIndex >= 0) {
                result.splice(lastDeletionIndex + 1, 0, current);
                // The new `lastDeletionIndex` may be out of range, but `splice` will fix it
                return [result, current, lastDeletionIndex + 2];
            }

            result.push(current);

            // Keep the `lastDeletionIndex` if there are lines of deletions,
            // otherwise update it to the new deletion line
            const newLastDeletionIndex = current.isDelete ? (last.isDelete ? lastDeletionIndex : i) : i;

            return [result, current, newLastDeletionIndex];
        },
        [[], null, -1]
    );
    return result;
};

const mapHunk = (hunk, options) => {
    const changes = options.nearbySequences === 'zip' ? zipChanges(hunk.changes) : hunk.changes;

    return {
        ...hunk,
        isPlain: false,
        changes: changes,
    };
};

const mapFile = (file, options) => {
    const hunks = file.hunks.map(hunk => mapHunk(hunk, options));

    return {...file, hunks};
};

const normalizeDiffText = text => {
    // Git diff header:
    //
    // diff --git a/test/fixture/test/ci.go b/test/fixture/test/ci.go
    // index 6829b8a2..4c565f1b 100644
    // --- a/test/fixture/test/ci.go
    // +++ b/test/fixture/test/ci.go
    if (text.indexOf('diff --git') === 0) {
        return text;
    }

    // Unidiff header:
    //
    // --- /test/fixture/test/ci.go 2002-02-21 23:30:39.942229878 -0800
    // +++ /test/fixture/test/ci.go 2002-02-21 23:30:50.442260588 -0800
    const indexOfFirstLineBreak = text.indexOf('\n');
    const indexOfSecondLineBreak = text.indexOf('\n', indexOfFirstLineBreak + 1);
    const firstLine = text.slice(0, indexOfFirstLineBreak);
    const secondLine = text.slice(indexOfFirstLineBreak + 1, indexOfSecondLineBreak);
    const oldPath = firstLine.split(' ').slice(1, -3).join(' ');
    const newPath = secondLine.split(' ').slice(1, -3).join(' ');
    const segments = [
        `diff --git a/${oldPath} b/${newPath}`,
        'index 1111111..2222222 100644',
        `--- a/${oldPath}`,
        `+++ b/${newPath}`,
        text.slice(indexOfSecondLineBreak + 1),
    ];

    return segments.join('\n');
};

export const parseDiff = (text, options = {}) => {
    const diffText = normalizeDiffText(text.trim());
    const files = parser.parse(diffText);

    return files.map(file => mapFile(file, options));
};
