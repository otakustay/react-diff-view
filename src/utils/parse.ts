import parser, {Change, DeleteChange, File, Hunk, InsertChange, NormalChange} from 'gitdiff-parser';

export function isInsert(change: Change): change is InsertChange {
    return change.type === 'insert';
}

export function isDelete(change: Change): change is DeleteChange {
    return change.type === 'delete';
}

export function isNormal(change: Change): change is NormalChange {
    return change.type === 'normal';
}

export type {File as FileData, Hunk as HunkData, Change as ChangeData};

export interface ParseOptions {
    nearbySequences?: 'zip';
}

function zipChanges(changes: Change[]) {
    const [result] = changes.reduce<[Change[], Change | null, number]>(
        ([result, last, lastDeletionIndex], current, i) => {
            if (!last) {
                result.push(current);
                return [result, current, isDelete(current) ? i : -1];
            }

            if (isInsert(current) && lastDeletionIndex >= 0) {
                result.splice(lastDeletionIndex + 1, 0, current);
                // The new `lastDeletionIndex` may be out of range, but `splice` will fix it
                return [result, current, lastDeletionIndex + 2];
            }

            result.push(current);

            // Keep the `lastDeletionIndex` if there are lines of deletions,
            // otherwise update it to the new deletion line
            const newLastDeletionIndex = isDelete(current) ? (isDelete(last) ? lastDeletionIndex : i) : i;

            return [result, current, newLastDeletionIndex];
        },
        [[], null, -1]
    );
    return result;
}

function mapHunk(hunk: Hunk, options: ParseOptions) {
    const changes = options.nearbySequences === 'zip' ? zipChanges(hunk.changes) : hunk.changes;

    return {
        ...hunk,
        isPlain: false,
        changes: changes,
    };
}

function mapFile(file: File, options: ParseOptions) {
    const hunks = file.hunks.map(hunk => mapHunk(hunk, options));

    return {...file, hunks};
}

function normalizeDiffText(text: string) {
    // Git diff header:
    //
    // diff --git a/test/fixture/test/ci.go b/test/fixture/test/ci.go
    // index 6829b8a2..4c565f1b 100644
    // --- a/test/fixture/test/ci.go
    // +++ b/test/fixture/test/ci.go
    if (text.startsWith('diff --git')) {
        return text;
    }

    // Unidiff header:
    //
    // --- /test/fixture/test/ci.go 2002-02-21 23:30:39.942229878 -0800
    // +++ /test/fixture/test/ci.go 2002-02-21 23:30:50.442260588 -0800
    const indexOfFirstLineBreak = text.indexOf('\n');
    const indexOfSecondLineBreak = text.indexOf('\n', indexOfFirstLineBreak + 1);
    const firstLine = text.slice(0, indexOfFirstLineBreak);
    const secondLine = text.slice(indexOfFirstLineBreak + 1, indexOfSecondLineBreak);
    const oldPath = firstLine.split(' ').slice(1, -3).join(' ');
    const newPath = secondLine.split(' ').slice(1, -3).join(' ');
    const segments = [
        `diff --git a/${oldPath} b/${newPath}`,
        'index 1111111..2222222 100644',
        `--- a/${oldPath}`,
        `+++ b/${newPath}`,
        text.slice(indexOfSecondLineBreak + 1),
    ];

    return segments.join('\n');
}

export function parseDiff(text: string, options: ParseOptions = {}): File[] {
    const diffText = normalizeDiffText(text.trim());
    const files = parser.parse(diffText);

    return files.map(file => mapFile(file, options));
}
