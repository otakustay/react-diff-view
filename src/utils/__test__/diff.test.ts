import {describe, test, expect} from 'vitest';
import dedent from 'dedent';
import {parseDiff} from '../parse';
import {
    textLinesToHunk,
    insertHunk,
    expandFromRawCode,
    getCollapsedLinesCountBetween,
    expandCollapsedBlockBy,
    getChangeKey,
    Change,
    Hunk,
} from '../index';

const sample = dedent`
    diff --git a/src/__test__/index.test.jsx b/src/__test__/index.test.jsx
    index 643c2f0..7883597 100644
    --- a/src/__test__/index.test.jsx
    +++ b/src/__test__/index.test.jsx
    @@ -21,3 +21,3 @@ describe('basic test', () => {
         test('App renders correctly', () => {
    -        expect(renderer.create(<App diffText={'deff'} />).toJSON()).toMatchSnapshot();
    +        expect(renderer.create(<App diffText={'diff'} />).toJSON()).toMatchSnapshot();
         });
`;

export const sampleHunk = parseDiff(sample)[0].hunks[0];

const normalChange: Change = {type: 'normal', isNormal: true, oldLineNumber: 1, newLineNumber: 1, content: ''};

const insertChange: Change = {type: 'insert', isInsert: true, lineNumber: 2, content: ''};

const deleteChange: Change = {type: 'delete', isDelete: true, lineNumber: 2, content: ''};

describe('textLinesToHunk', () => {
    test('basic', () => {
        const lines = [''];
        expect(textLinesToHunk(lines, 0, 0)).toMatchSnapshot();
    });
});

describe('insertHunk', () => {
    test('basic', () => {
        const results = insertHunk(
            [{changes: [normalChange], content: '', oldStart: 1, newStart: 1, oldLines: 1, newLines: 1}],
            {changes: [deleteChange], content: '', oldStart: 2, newStart: 2, oldLines: 1, newLines: 1}
        );
        expect(results.length).toBe(1);
        expect(results[0].changes).toEqual([normalChange, deleteChange]);
    });
});

describe('expandFromRawCode', () => {
    test('basic', () => {
        const hunks = [sampleHunk];
        expect(expandFromRawCode(hunks, '', 22, 23)).toMatchSnapshot();
    });
});

describe('getCollapsedLinesCountBetween', () => {
    test('basic', () => {
        const previousHunk: Hunk = {content: '', oldStart: 1, oldLines: 2, newStart: 1, newLines: 2, changes: []};
        const nextHunk: Hunk = {content: '', oldStart: 10, oldLines: 2, newStart: 10, newLines: 2, changes: []};
        expect(getCollapsedLinesCountBetween(previousHunk, nextHunk)).toBe(7);
    });

    test('minus number', () => {
        const previousHunk: Hunk = {content: '', oldStart: 1, oldLines: 10, newStart: 1, newLines: 2, changes: []};
        const nextHunk: Hunk = {content: '', oldStart: 2, oldLines: 2, newStart: 10, newLines: 2, changes: []};
        expect(getCollapsedLinesCountBetween(previousHunk, nextHunk)).toBe(-9);
    });

    test('no previousHunk', () => {
        const nextHunk: Hunk = {content: '', oldStart: 2, oldLines: 2, newStart: 10, newLines: 2, changes: []};
        expect(getCollapsedLinesCountBetween(null, nextHunk)).toBe(1);
    });
});

describe('expandCollapsedBlockBy', () => {
    test('basic', () => {
        const hunks = [sampleHunk];
        expect(expandCollapsedBlockBy(hunks, '', () => false)).toMatchSnapshot();
    });

    test('normal hunk', () => {
        const hunks: Hunk[] = [
            {
                content: '@@ -1,2 +1,2 @@',
                oldStart: 1,
                newStart: 1,
                oldLines: 2,
                newLines: 2,
                changes: [{
                    content: 'iiiiiiiiiiiiiiiiiiiiii:WQiiiiiiiiiiiiejj',
                    type: 'normal',
                    isNormal: true,
                    oldLineNumber: 1,
                    newLineNumber: 1,
                }, {
                    content: 'dsds',
                    type: 'delete',
                    isDelete: true,
                    lineNumber: 2,
                }, {
                    content: 'dsdsds',
                    type: 'insert',
                    isInsert: true,
                    lineNumber: 2,
                }],
            }
        ];
        const rawCode = 'iiiiiiiiiiiiiiiiiiiiii:WQiiiiiiiiiiiiejj\ndsds';
        expect(expandCollapsedBlockBy(hunks, rawCode, () => false)).toMatchSnapshot();
        expect(expandCollapsedBlockBy(hunks, rawCode.split('\n'), () => false)).toMatchSnapshot();
        expect(expandFromRawCode(hunks, rawCode, 0, 10)).toMatchSnapshot();
    });
});

describe('getChangeKey', () => {
    test('normal change', () => {
        expect(getChangeKey(normalChange)).toBe('N1');
    });

    test('insert change', () => {
        expect(getChangeKey(insertChange)).toBe('I2');
    });

    test('delete change', () => {
        expect(getChangeKey(deleteChange)).toBe('D2');
    });
});
