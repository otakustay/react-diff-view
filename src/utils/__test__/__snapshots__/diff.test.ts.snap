// Vitest Snapshot v1

exports[`expandCollapsedBlockBy > basic 1`] = `
[
  {
    "changes": [
      {
        "content": "    test('App renders correctly', () => {",
        "isNormal": true,
        "newLineNumber": 21,
        "oldLineNumber": 21,
        "type": "normal",
      },
      {
        "content": "        expect(renderer.create(<App diffText={'deff'} />).toJSON()).toMatchSnapshot();",
        "isDelete": true,
        "lineNumber": 22,
        "type": "delete",
      },
      {
        "content": "        expect(renderer.create(<App diffText={'diff'} />).toJSON()).toMatchSnapshot();",
        "isInsert": true,
        "lineNumber": 22,
        "type": "insert",
      },
      {
        "content": "    });",
        "isNormal": true,
        "newLineNumber": 23,
        "oldLineNumber": 23,
        "type": "normal",
      },
    ],
    "content": "@@ -21,3 +21,3 @@ describe('basic test', () => {",
    "isPlain": false,
    "newLines": 3,
    "newStart": 21,
    "oldLines": 3,
    "oldStart": 21,
  },
]
`;

exports[`expandCollapsedBlockBy > normal hunk 1`] = `
[
  {
    "changes": [
      {
        "content": "iiiiiiiiiiiiiiiiiiiiii:WQiiiiiiiiiiiiejj",
        "isNormal": true,
        "newLineNumber": 1,
        "oldLineNumber": 1,
        "type": "normal",
      },
      {
        "content": "dsds",
        "isDelete": true,
        "lineNumber": 2,
        "type": "delete",
      },
      {
        "content": "dsdsds",
        "isInsert": true,
        "lineNumber": 2,
        "type": "insert",
      },
    ],
    "content": "@@ -1,2 +1,2 @@",
    "newLines": 2,
    "newStart": 1,
    "oldLines": 2,
    "oldStart": 1,
  },
]
`;

exports[`expandCollapsedBlockBy > normal hunk 2`] = `
[
  {
    "changes": [
      {
        "content": "iiiiiiiiiiiiiiiiiiiiii:WQiiiiiiiiiiiiejj",
        "isNormal": true,
        "newLineNumber": 1,
        "oldLineNumber": 1,
        "type": "normal",
      },
      {
        "content": "dsds",
        "isDelete": true,
        "lineNumber": 2,
        "type": "delete",
      },
      {
        "content": "dsdsds",
        "isInsert": true,
        "lineNumber": 2,
        "type": "insert",
      },
    ],
    "content": "@@ -1,2 +1,2 @@",
    "newLines": 2,
    "newStart": 1,
    "oldLines": 2,
    "oldStart": 1,
  },
]
`;

exports[`expandCollapsedBlockBy > normal hunk 3`] = `
[
  {
    "changes": [
      {
        "content": "iiiiiiiiiiiiiiiiiiiiii:WQiiiiiiiiiiiiejj",
        "isNormal": true,
        "newLineNumber": 0,
        "oldLineNumber": 0,
        "type": "normal",
      },
      {
        "content": "iiiiiiiiiiiiiiiiiiiiii:WQiiiiiiiiiiiiejj",
        "isNormal": true,
        "newLineNumber": 1,
        "oldLineNumber": 1,
        "type": "normal",
      },
      {
        "content": "dsds",
        "isDelete": true,
        "lineNumber": 2,
        "type": "delete",
      },
      {
        "content": "dsdsds",
        "isInsert": true,
        "lineNumber": 2,
        "type": "insert",
      },
    ],
    "content": "@@ -0,3 +0,3",
    "isPlain": false,
    "newLines": 3,
    "newStart": 0,
    "oldLines": 3,
    "oldStart": 0,
  },
]
`;

exports[`expandFromRawCode > basic 1`] = `
[
  {
    "changes": [
      {
        "content": "    test('App renders correctly', () => {",
        "isNormal": true,
        "newLineNumber": 21,
        "oldLineNumber": 21,
        "type": "normal",
      },
      {
        "content": "        expect(renderer.create(<App diffText={'deff'} />).toJSON()).toMatchSnapshot();",
        "isDelete": true,
        "lineNumber": 22,
        "type": "delete",
      },
      {
        "content": "        expect(renderer.create(<App diffText={'diff'} />).toJSON()).toMatchSnapshot();",
        "isInsert": true,
        "lineNumber": 22,
        "type": "insert",
      },
      {
        "content": "    });",
        "isNormal": true,
        "newLineNumber": 23,
        "oldLineNumber": 23,
        "type": "normal",
      },
    ],
    "content": "@@ -21,3 +21,3 @@ describe('basic test', () => {",
    "isPlain": false,
    "newLines": 3,
    "newStart": 21,
    "oldLines": 3,
    "oldStart": 21,
  },
]
`;

exports[`textLinesToHunk > basic 1`] = `
{
  "changes": [
    {
      "content": "",
      "isNormal": true,
      "newLineNumber": 0,
      "oldLineNumber": 0,
      "type": "normal",
    },
  ],
  "content": "@@ -0,1 +0,1",
  "isPlain": true,
  "newLines": 1,
  "newStart": 0,
  "oldLines": 1,
  "oldStart": 0,
}
`;
